src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -2952,6 +2953,174 @@ theorem sub_le_sub_iff_le {x y z : BitVec w} (hxz : z ≤ x) (hyz : z ≤ y) :
+    carry (i+1) x (1#w) false = decide (∀ j ≤ i, x.getLsbD j = true) := by
+  sorry
+
+/--
+Remember that a negating a bitvector is equal to incrementing the complement
+by one, i.e., `-x = ~~~x + 1`. (see `neg_eq_not_add`)
+
+Thus, if the least-significant-bit of `x` is `true`, then its complement is
+`false`, we add one to it, and we get that the lsb of the negation is `true`,
+and the carry disappeared, so the rest of the bits remain complemented.
+
+Otherwise, if the least-significant bit of `x` was `false`, then its complement
+is `true`, which flips back to the original value of `false` after adding one,
+and the carry remains.
+
+Extrapolating, we find that the `i`-th bit of `-x` is equal to the `i`-th bit
+of the input `x` iff the least significant `i` bits of `x` (including the `i`-th
+bit itself) are `false`.
+Otherwise, if one of those bits is `true`, the `i`-th bit of `-x` is the
+complement of the `i`-th bit of `x`.
+-/
+theorem getLsbD_neg {i : Nat} {x : BitVec w} :
+    getLsbD (-x) i =
+      (getLsbD x i ^^ decide (i < w ∧ ∃ j < i, getLsbD x j = true)) := by
+  rw [neg_eq_not_add]
+  by_cases hi : i < w
+  case neg =>
+    have h_ge : w ≤ i := by omega
+    simp [getLsbD_ge _ _ h_ge, h_ge, hi]
+  case pos =>
+    rw [getLsbD_add hi]
+    have : 0 < w := by omega
+    simp only [getLsbD_not, hi, decide_True, Bool.true_and, getLsbD_one, this, Bool.not_bne,
+      Bool.not_eq_eq_eq_not, true_and, bne_iff_ne, ne_eq, decide_eq_decide]
+
+    cases i
+    case zero =>
+      have carry_zero : carry 0 ?x ?y false = false := by
+          simp [carry]; omega
+      simp [hi, carry_zero]
+    case succ i =>
+      rw [carry_incr _ _ (by omega), ← Bool.xor_not, ← decide_not]
+      simp only [Nat.add_one_ne_zero, decide_False, getLsbD_not, Bool.and_eq_true,
+        decide_eq_true_eq, Bool.not_eq_eq_eq_not, Bool.not_true, Bool.false_bne, not_exists,
+        not_and, Bool.not_eq_true, Bool.bne_left_inj, decide_eq_decide]
+      constructor
+      · rintro h j hj; apply And.right <| h j (by omega)
+      · rintro h j hj; exact ⟨by omega, h j (by omega)⟩
+
+theorem getMsbD_neg {i : Nat} {x : BitVec w} :
+    getMsbD (-x) i =
+      (getMsbD x i ^^ decide (∃ j < w, i < j ∧ getMsbD x j = true)) := by
+  simp only [getMsbD, getLsbD_neg, Bool.decide_and, Bool.and_eq_true, decide_eq_true_eq]
+  by_cases hi : i < w
+  case neg =>
+    simp [hi]; omega
+  case pos =>
+    have h₁ : w - 1 - i < w := by omega
+    simp only [hi, decide_True, h₁, Bool.true_and, Bool.bne_left_inj, decide_eq_decide]
+    constructor
+    · rintro ⟨j, hj, h⟩
+      refine ⟨w - 1 - j, ?_, ?_, ?_, _root_.cast ?_ h⟩
+      <;> try omega
+      congr; omega
+    · rintro ⟨j, hj₁, hj₂, -, h⟩
+      refine ⟨w - 1 - j, ?_, h⟩
+      omega
+
+theorem msb_neg {w : Nat} {x : BitVec w} :
+    (-x).msb = ((!decide (x = 0#w) && !decide (x = intMin w)) ^^ x.msb) := by
+  have getMsbD_intMin {w i} : (intMin w).getMsbD i = decide (0 < w ∧ i = 0) :=
+      sorry
```

Comment:
And here's another sorry, but this one shouldn't be too hard. I think we had another PR floating around that proved intMin results, so I didn't even attempt this lemma